docs: add The State of Flint - Spring 2026 blog post (#2433)
flint-fyi/flint: packages/site/src/content/docs/blog/the-state-of-flint-2026-spring.mdx

> - [What Flint Does Differently > Interactive and Watch CLI Modes](/blog/what-flint-does-differently#interactive-and-watch-cli-modes)

### Lint Rule Progress

The Flint project encapsulates more than just a solid linter foundation and CLI.
It's also meant to explore a having a cohesive, comprehensive set of lint rules applicable to modern users of the linter.
We expect to end up with roughly 1,000 rules implemented in the Flint monorepo, and roughly half of those rules to be in Flint's TypeScript plugin.

As of Spring 2026, the Flint monorepo has implemented:

- Approximately 400 of all 1,000 rules (~40% of the goal)
- Of those, approximately 300 of the TypeScript rules (~60% of the goal)

Flint's implemented rules intentionally span the gamut of linting checks.
They include behavior from the core ESLint and typescript-eslint projects, plugins such as [eslint-plugin-regexp](https://ota-meshi.github.io/eslint-plugin-regexp), and languages such as JSON and Markdown.

> [What Flint Does Differently > Common Core Rules](/blog/what-flint-does-differently#common-core-rules)

### Volar and Vue Integrations

> Flint's goal of best-in-class, fully-featured linting necessitates first-class support for "extension" languages such as Astro, Svelte, and Vue.

Author: I think we should delay that deadline by a few days because I took a few days to review 😅 sorry.

It also provides low-level Node.js APIs for running linting on one or more files.

`@flint.fyi/core` is "initially" feature-complete as of Spring 2026.
The main pieces of functionality all work as expected.
While we expect to continue to spot iterate on it to support future features, we have no major known work items in the backlog for it that block known future work.

#### Cross-File Linting

Flint's biggest core piece of advancement is first-class support for "cross-file" linting: rules that may pull in information from multiple files at a time.
The most common form of cross-file linting is typed linting, which is linting that uses type information.
Flint provides full native support for cross-file linting:

- Its cache factors in cross-file dependencies such as imports and `tsconfig.json` files
- Rule suggestions may apply changes across files, not just to the file being linted
- Rules are given "setup" and "teardown" hooks to support stateful linting logic

Flint already provides several core rules that make use of these features in ways most or all other linters do not.

> Its experimental spell-checking rule, for example, provides suggestions to add unknown words to the user's dictionary file.

Reviewer: Maybe I'm being pedantic here, but do we have any rules that close over multiple files? The cspell fix is a hack right now and doesn't really use the cross-file linting architecture (yes it uses the teardown, but just to orchestrate I/O work). Now that I say it, do we of yet have any rules that close over the environment to provide unique insights beyond TypeScript types?

> - [What Flint Does Differently > Cross File Caching](/blog/what-flint-does-differently#cross-file-caching)
> - [What Flint Does Differently > Rich Cross File Fixes](/blog/what-flint-does-differently#rich-cross-file-fixes)

#### Tooling Coordination

Another banner feature for Flint is running formatting and type-checking alongside the linter.
Doing so can reduce tooling configuration complexity, as well as reduce the amount of time (and therefore money) spent on CI.

This has been implemented in Flint.
Flint runs Prettier on all linted files after linting, piping through its own `--fix` flag to format files if requested.

> Flint also provides languages with API hooks to define their own "diagnostics" that can be reported.

Reviewer 2: I would *really* love to find a different name for diagnostics.

Reviewer: Cross linking #1773

The TypeScript language uses those hooks to reuse type information services from linting and report type errors.

> - [What Flint Does Differently > Formatting Coordination](/blog/what-flint-does-differently#formatting-coordination)
> - [What Flint Does Differently > Language Diagnostics Coordination](/blog/what-flint-does-differently#language-diagnostics-coordination)

#### Virtual File System Host

One notable part of `@flint.fyi/core` is a full virtual file system ("VFS" host).
Meaning: Flint provides an isolated in-memory representation of directories and files for linting to operate on.
Flint's VFS host defaults to the local file system, but can also be used in other environments.

Establishing a VFS host early will allow Flint to create well-architected APIs for environments such as in-memory unit tests and web browsers.

> It also allows for a number of performance optimizations, such as the ability to batch file system reads and writes.

Reviewer: I think the important part of the Host is that _all_ I/O is abstracted out, so you can use it for a VFS, but you can also just do whatever you want with Flint's internals.
Also a few wording tweaks to go along with that.

I'm sure there's a better name than I/O Host but VFS Host is underselling it.

```suggestion
#### I/O Host

One notable part of `@flint.fyi/core` is that we use a virtual file system underlying our I/O host.
This means that Flint provides an isolated in-memory representation of directories and files for lints to operate on.
Flint's VFS host defaults to the local file system, but can also be used in other environments such as in-memory unit tests.

Establishing host architecture early will allow Flint to support environments like isolated tests and [the web](#playground-inspector).
It also allows for a number of (yet-to-be-implemented) performance optimizations, such as the ability to batch file system reads and writes.
```

> [What Flint Does Differently > Embeddable By Design](/blog/what-flint-does-differently#embeddable-by-design)

### CLI

> The Flint CLI, or `npx flint`, is the primary -and currently only- way to run Flint.

Reviewer: I don't love calling us `npx flint`. Maybe just `flint`?

```suggestion
The Flint CLI, or `flint`, is the primary—and currently only—way to run Flint.
```

Author: Yeah I like just `flint`. More and more I've started to use things like `pnpx`...

!["DROP THE 'THE'" scene clip from The Social Network](https://github.com/user-attachments/assets/2f1623f4-9f9b-4cd7-a138-fa6ff0f848e0)

> It's intentionally analagous to existing lint CLIs such as `npx biome`, `npx eslint`, and `npx oxlit`.

Reviewer 3: ```suggestion
It's intentionally analagous to existing lint CLIs such as `npx biome`, `npx eslint`, and `npx oxlint`.
```

`npx flint` is largely feature-complete as of Spring 2026.
It supports:

- Reading rules from a `flint.config.*` file
- Reading past lint results from cache, if available
- Detecting which files were changed since the cache was created
- Running linting on changed files, or all if no cache was available
- `--fix` to auto-fix reported violations from rules
- `--fix-suggestions` to apply specific unsafe lint rule suggestions as fixes
- `--interactive` to run in a keyboard-controlled one-at-a-time view
- `--presenter brief|detailed` to control how many details are shown with lint rule reports
- `--watch` mode to re-run on files as they save changes

In fact, we've been running Flint in CI alongside ESLint and other checks for the majority of the Flint project.

> We've found it to be stable.

Author: I think a more actionable, tangible thing here would be to say that we could start trying it out on our own projects.

@barrymichaeldoyle and I were actually talking about this: finding some real-world projects separate from Flint to try out!

> - [What Flint Does Differently > Comprehensive Rule Reports](/blog/what-flint-does-differently#comprehensive-rule-reports)
> - [What Flint Does Differently > Interactive and Watch CLI Modes](/blog/what-flint-does-differently#interactive-and-watch-cli-modes)

### Lint Rule Progress

The Flint project encapsulates more than just a solid linter foundation and CLI.

> It's also meant to explore a having a cohesive, comprehensive set of lint rules applicable to modern users of the linter.

Reviewer 2: I think there's an extra "a" here.

These users range from complete linting novices to power users configuring their individual rule options.
For them, we'll have a glossary, FAQs, introductory docs, and guided walkthroughs of common tasks.
- **Plugin developer**: showing how to use Flint's APIs to work with custom plugins and write custom rules.
Man of these users are writing small custom rules to achieve company/project-specific linting tasks.
These developers will have reference documentation and sample projects showing how to use Flint.
- **Linter developer**: someone working on the linter itself, or at least directly using its programmatic APIs.
These users often want to know how the linter is architected at a high-level.
For them, we plan to have full architecture documentation, Node.js API reference pages, and guides for basic usage of the API.

The glossary and FAQs were started in 2025.
We plan on fleshing out most or all the rest of that documentation as the project solidifies its architecture and best practices through 2026.

> [What Flint Does Differently > Shared Glossary](/blog/what-flint-does-differently#shared-glossary)

> ### Editor Integrations (LSP)

Reviewer 3: Can't find the right words (and I'm not sure whether to put it under "Editor Integrations" or under "Virtual File System Host"), but we can probably mention that our languages are now stateful thanks to the `LinterHost`, so `@flint.fyi/typescript-language` will be able to provide nice DX since it's able to operate on multiple not-yet-saved files opened in the text editor, and so on...

Most developers dislike seeing angry red squigglies in their editor.
But receiving instant feedback on changes, with triggerable actions such as applying auto-fixes, is a core part of the linting experience users expect.
Ironic.

Flint will provide a [Language Server Protocol ("LSP")](https://en.wikipedia.org/wiki/Language_Server_Protocol) server as well as at least a VS Code extension built atop it.
Doing so will enable most users of the linter to integrate Flint into their editor.
Now that Flint's core has been mostly stable for several months, we're looking forward to building up higher-level APIs that can power LSP usage.

### MCP Server

> If an LSP server is de facto required for humans in editors, then a [Model Context Protocol ("MCP")](https://en.wikipedia.org/wiki/Model_Context_Protocol) is de facto required for agentic workflows.

Author: Well, I made it in PST time.

AIs are steadily growing in programmer popularity and power.
Setting up an MCP interface so the linter can directly guide agents through fixing code flaws will be table stakes for any linter soon.

As with LSP, we're looking forward to building up higher-level Flint APIs that can power MCP usage.

### More Lint Rules

Flint's lint rules goal is to encapsulate roughly _all_ important lint rules common users will want.
That will take time.

Our plan is to continuously add more rules each month, with milestones every two months of development.
Our optimistic timeline for rule completions will get to ~700 lint rules by the end of 2026.
Starting with ~400 implemented rules in Spring 2026, that projection is:

1. End of April: ~510 (+110)
1. End of June: ~585 (+75)
1. End of August: ~635 (+50)
1. End of October: ~670 (+35)
1. End of December: ~700 (+30)

We expect the number of rules added each month to go down as easier rules are tackled first.

> For example, most of the rules we expect to add from [eslint-plugin-package-json](https://github.com/michaelfaith/eslint-plugin-package-json) in April are straightforward or even auto-generated from schemas.

Reviewer 2: I've been meaning to come back to the issue where it was being discussed, but did structural visitor queries ever get implemented?  E.g. 'root > ObjectExpression > Property[name="exports"]'? 

We use that pretty exclusively in eppj for property specific rules (which is most of them).

Reviewer 2: It was mentioned in the original issue description too, just buried a bit

> Right now, Flint's visitors don't support ESQuery-style query selectors like VariableDeclaration > Identifier or VariableDeclaration:exit

My comment was directly responding to that piece.  Just calling it out because it's going to make re-implementing the package-json rule a bit more tedious.

Author: Oh yeah this whole thing. I'm not hard-set against them but there are definite downsides -as auvred noted- that I think we should discuss if we want to look into them. That's why I didn't implement them in 2025 and haven't pushed for them since. But if someone wants to file an issue I'm down to think through things more and post reasoning!

> [What Flint Does Differently > Common Core Rules](/blog/what-flint-does-differently#common-core-rules)

:::tip
Expect a blog post later this year on how the Flint project is using a combination of AI and manual review to create dozens of high quality lint rule (re-)implementations a month.
:::

### Performance Deep Dives

Flint is still wholly written in TypeScript and does not directly farm out to any native code.
That means it not yet truly a "hybrid linter" yet.
All of Flint's performance work so far has been focused on making its core implemented-in-TypeScript parts fast and lean.

Implementing hybrid linting is a daunting, difficult task.

> No project has implemented quite the level of Go-to-TypeScript communication that Flint will need to work with typescript-go.

Reviewer: Shhhhh 🤫 <sub><sup><sub><sup><sub><sup>_Golar_ 😉</sup></sub></sup></sub></sup></sub> (jk. Maybe?)

However, as with configuration projects / workspaces, this feature is not essential for a pre-alpha linter.
We don't yet have users who will need it.
A configuration inspector will be a high-value add for Flint when it gets closer a userland beta phase in 2027 or later.

### ESLint Compatibility Layer

ESLint is the predominant linter in web developer tooling today.
Many users of ESLint who might want to try out an alternate linter use community plugins or even their own custom rules that won't be available in Flint.
Providing a compatibility layer to run those rules in Flint will be essential for letting those users try out Flint.

However, as with configuration projects / workspaces and the configuration inspector, this feature is not essential for a pre-alpha linter.
We don't yet have users who will need it.
An ESLint compatibility layer will be a high-value add for Flint when it gets closer a userland beta phase in 2027 or later.

### Playground Inspector

Every other mainstream linter provides an online "playground" where you can preview individual rules on small snippets of source code.
These playgrounds are immensely helpful for understanding linters and their rules.

> They're also integrated into the GitHub flow, such as playground reproductions being required for many basic forms of issues.

Author: Oof, yeah, now that you mention it...

  For this first post, we'll cover progress made from Flint's start in 2025 and through first few months of 2026.
title: "The State of Flint - Spring 2026"
---

import { Aside } from "@astrojs/starlight/components";

Hello, and welcome to the first _State of Flint_ blog post!
These posts explain the high-level status of the Flint project: what's been done, what's coming up next, and what's being pushed to the backlog.
We plan on publishing these posts 1-2 times a year as the Flint project hits its major milestones.

:::tip
New to Flint?
See [About Flint](/about) to learn more about the fast, friendly linter for JavaScript, TypeScript, and more.
:::

## What We've Done

For this first post, we'll cover progress made from Flint's start in 2025 and through first few months of 2026.

### Flint Core

Flint's "Core" package, `@flint.fyi/core`, contains the main primitives for setting up a linter runtime, defining plugins for that runtime, and creating rules for those plugins.
It also provides low-level Node.js APIs for running linting on one or more files.

> `@flint.fyi/core` is, compared to other linters, feature-complete as of Spring 2026.

Reviewer 2: Hmm, I think it lost that important asterisk though (i.e. feature complete when you're comparing it to other linters).  It might give the idea that the project is further along than it is?  Maybe something like this?

```suggestion
When it comes to the offerings of most traditional linters, `@flint.fyi/core` is feature-complete as of Spring 2026.
```

### MCP Server

If an LSP server is de facto required for humans in editors, then many developers have stated a [Model Context Protocol ("MCP")](https://en.wikipedia.org/wiki/Model_Context_Protocol) server to be de facto required for agentic workflows.
AI is steadily growing in programmer popularity and power.

> Figuring out the right way for AI to interact with Flint so the linter can directly guide agents through fixing code flaws will be table stakes for any linter soon.

Reviewer 4: With the latest edit, this sentence starts out talking about "Flint" specifically, but ends generically on "any linter".
